File: 3031-construct-product-matrix/construct-product-matrix.py
class Solution:
    def constructProductMatrix(self, grid: List[List[int]]) -> List[List[int]]:
        MOD = 12345
        n = len(grid)
        m = len(grid[0])
        p = [[0] * m for _ in range(n)]

        suffix = 1
        for i in range(n - 1, -1, -1):
            for j in range(m - 1, -1, -1):
                p[i][j] = suffix
                suffix = (suffix * grid[i][j]) % MOD

        prefix = 1
        for i in range(n):
            for j in range(m):
                p[i][j] = (prefix * p[i][j]) % MOD
                prefix = (prefix * grid[i][j]) % MOD

        return p

File: 3031-construct-product-matrix/construct-product-matrix.js
/**
 * @param {number[][]} grid
 * @return {number[][]}
 */
var constructProductMatrix = function (grid) {
    const MOD = 12345;
    const n = grid.length;
    const m = grid[0].length;
    const p = Array.from({ length: n }, () => Array(m).fill(0));

    let suffix = 1n;
    for (let i = n - 1; i >= 0; i--) {
        for (let j = m - 1; j >= 0; j--) {
            p[i][j] = Number(suffix);
            suffix = (suffix * BigInt(grid[i][j])) % BigInt(MOD);
        }
    }

    let prefix = 1n;
    for (let i = 0; i < n; i++) {
        for (let j = 0; j < m; j++) {
            p[i][j] = Number((prefix * BigInt(p[i][j])) % BigInt(MOD));
            prefix = (prefix * BigInt(grid[i][j])) % BigInt(MOD);
        }
    }
    return p;
};